tests/unit/api/objectPutVersioned.js

```diff
@@ -0,0 +1,199 @@
+import async from 'async';
+import assert from 'assert';
+
+import bucketPut from '../../../lib/api/bucketPut';
+import bucketPutVersioning from '../../../lib/api/bucketPutVersioning';
+import objectPut from '../../../lib/api/objectPut';
+import { ds } from '../../../lib/data/in_memory/backend';
+import DummyRequest from '../DummyRequest';
+import { cleanup, DummyRequestLogger, makeAuthInfo } from '../helpers';
+
+const log = new DummyRequestLogger();
+const canonicalID = 'accessKey1';
+const authInfo = makeAuthInfo(canonicalID);
+const namespace = 'default';
+const bucketName = 'bucketname';
+const objectName = 'objectName';
+
+const testPutBucketRequest = new DummyRequest({
+    bucketName,
+    namespace,
+    headers: { host: `${bucketName}.s3.amazonaws.com` },
+    url: '/',
+});
+
+function _createBucketPutVersioningReq(status) {
+    const request = {
+        bucketName,
+        headers: {
+            host: `${bucketName}.s3.amazonaws.com`,
+        },
+        url: '/?versioning',
+        query: { versioning: '' },
+    };
+    const xml = '<VersioningConfiguration ' +
+    'xmlns="http://s3.amazonaws.com/doc/2006-03-01/">' +
+    `<Status>${status}</Status>` +
+    '</VersioningConfiguration>';
+    request.post = xml;
+    return request;
+}
+
+function _createPutObjectRequest(body) {
+    const params = {
+        bucketName,
+        namespace,
+        objectKey: objectName,
+        headers: {},
+        url: `/${bucketName}/${objectName}`,
+    };
+    return new DummyRequest(params, body);
+}
+
+const enableVersioningRequest = _createBucketPutVersioningReq('Enabled');
+const suspendVersioningRequest = _createBucketPutVersioningReq('Suspended');
+
+describe('objectPut API with versioning', () => {
+    beforeEach(() => {
+        cleanup();
+    });
+
+    function _assertDataStoreValues(expectedValues) {
+        assert.strictEqual(ds.length, expectedValues.length + 1);
+        for (let i = 0, j = 1; i < expectedValues.length; i++, j++) {
+            if (expectedValues[i] === undefined) {
+                assert.strictEqual(ds[j], expectedValues[i]);
+            } else {
+                assert.deepStrictEqual(ds[j].value, expectedValues[i]);
+            }
+        }
+    }
+
+    it('should delete latest version when creating new null version ' +
+    'if latest version is null version', done => {
+        let i = 0;
```

Comment:
I think these tests would be clearer if you did them without 'i'.  You have a set array in each, just use the actual index you are trying to pull.